programs/server/HTTPHandlerFactory.cpp

```diff
@@ -40,4 +60,104 @@ Poco::Net::HTTPRequestHandler * HTTPRequestHandlerFactoryMain::createRequestHand
     return nullptr;
 }
 
+HTTPRequestHandlerFactoryMain::~HTTPRequestHandlerFactoryMain()
+{
+    while (!child_factories.empty())
+        delete child_factories.back(), child_factories.pop_back();
+}
+
+HTTPRequestHandlerFactoryMain::TThis * HTTPRequestHandlerFactoryMain::addHandler(Poco::Net::HTTPRequestHandlerFactory * child_factory)
+{
+    child_factories.emplace_back(child_factory);
+    return this;
+}
+
+static inline auto createHandlersFactoryFromConfig(IServer & server, const std::string & name, const String & prefix)
+{
+    auto main_handler_factory = new HTTPRequestHandlerFactoryMain(name);
+
+    try
+    {
+        Poco::Util::AbstractConfiguration::Keys keys;
+        server.config().keys(prefix, keys);
+
+        for (const auto & key : keys)
+        {
+            if (!startsWith(key, "routing_rule"))
+                throw Exception("Unknown element in config: " + prefix + "." + key + ", must be 'routing_rule'", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+
+            const auto & handler_type = server.config().getString(prefix + "." + key + ".handler.type", "");
+
+            if (handler_type == "static")
+                main_handler_factory->addHandler(createStaticHandlerFactory(server, prefix + "." + key));
+            else if (handler_type == "dynamic_query_handler")
+                main_handler_factory->addHandler(createDynamicHandlerFactory(server, prefix + "." + key));
+            else if (handler_type == "predefine_query_handler")
+                main_handler_factory->addHandler(createPredefineHandlerFactory(server, prefix + "." + key));
+            else
+                throw Exception("Unknown element in config: " + prefix + "." + key + ", must be 'routing_rule'", ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
+        }
+
+        return main_handler_factory;
+    }
+    catch (...)
+    {
+        delete main_handler_factory;
+        throw;
+    }
+}
+
+static const auto ping_response_expression = "Ok.\n";
+static const auto root_response_expression = "config://http_server_default_response";
+
+static inline Poco::Net::HTTPRequestHandlerFactory * createHTTPHandlerFactory(IServer & server, const std::string & name, AsynchronousMetrics & async_metrics)
+{
+    if (server.config().has("routing_rules"))
+        return createHandlersFactoryFromConfig(server, name, "routing_rules");
```

Comment:
Is this the intention that `http_handlers` (it was renamed in merging PR - #10547) overrides default paths? i.e. /ping, /replicas_status
Maybe it will better to use separate server for custom http handlers (on another port)? Or register custom http handlers first and then defaults, to disallow their overrides (or allow overrides, but register anyway)

Reply:
@zhang2014 @tavplubix thoughts?